File: 02_Stack/03_Evaluate_Reverse_Polish_Notation/150_Evaluate_Reverse_Polish_Notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> s;

        for (string token : tokens){
            if (token == "+"){
                int l = s.top();
                s.pop();
                int r = s.top();
                s.pop();
                s.push(l+r);
            }
            else if (token == "-"){
                int l = s.top();
                s.pop();
                int r = s.top();
                s.pop();
                s.push(r-l);
            }
            else if (token == "*"){
                int l = s.top();
                s.pop();
                int r = s.top();
                s.pop();
                s.push(r*l);
            }
            else if (token == "/"){
                int l = s.top();
                s.pop();
                int r = s.top();
                s.pop();
                s.push(r/l);
            }
            else {
                s.push(stoi(token));
            }
        }
        return s.top();
    }
};

File: 02_Stack/03_Evaluate_Reverse_Polish_Notation/150_Evaluate_Reverse_Polish_Notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for e in tokens:
            match e:
                case "+":
                    r = stack.pop()
                    l = stack.pop()
                    stack.append(l+r)
                case "*":
                    r = stack.pop()
                    l = stack.pop()
                    stack.append(l*r)
                case "-":
                    r = stack.pop()
                    l = stack.pop()
                    stack.append(l-r)
                case "/":
                    r = stack.pop()
                    l = stack.pop()
                    if l == 0:
                        stack.append(0)
                    else:
                        stack.append(int(l / r))                    
                case _:
                    stack.append(int(e))
        
        return stack[0]